File: Code/06_circadian_resynchronization/matlab/circadian_recovery.m
clear; clc; close all;
addpath('../../common/matlab');
figDir = fullfile('..','..','..','Report','Figs');
twilightCols = twilight_colors(4);
blue = twilightCols(1,:); green = twilightCols(2,:);
red = twilightCols(3,:); magenta = twilightCols(4,:);
lightGray = [0.85 0.87 0.89]; black = [0.05 0.05 0.05];

%% Figure 15: phase-space dynamics
Delta = 1.0; F = 3.5 * Delta; Omega = 1.4 * Delta;
regimes = [10.0, 4.5] * Delta;
regimeLabels = {'K=10\Delta', 'K=4.5\Delta'};
shifts = [-8.5, -9.5, 9.0, 12.0];
shiftLabels = {'8.5 h eastward','9.5 h eastward','9.0 h westward','12 h E/W'};
shiftColors = {green, red, blue, magenta};
fig = figure('Position', [100 100 720 335], 'Color', 'w');
theta = linspace(0, 2*pi, 500);
for p = 1:2
    K = regimes(p);
    [zst, pts, types] = stable_fixed_point(K, F, Delta, Omega);
    subplot(1,2,p); hold on; axis equal;
    plot(cos(theta), sin(theta), '--', 'Color', lightGray, 'LineWidth', 1.0);
    plot([-1 1], [0 0], 'Color', lightGray, 'LineWidth', 0.9);
    plot([0 0], [-1 1], 'Color', lightGray, 'LineWidth', 0.9);
    for s = 1:numel(shifts)
        z0 = zst * exp(1i * shifts(s) * 2*pi/24);
        [~, ztraj] = simulate_traj(z0, [0 80], K, F, Delta, Omega);
        plot(real(ztraj), imag(ztraj), 'Color', shiftColors{s}, 'LineWidth', 1.5, 'DisplayName', shiftLabels{s});
        plot(real(ztraj(1)), imag(ztraj(1)), 'o', 'MarkerSize', 3.2, 'MarkerFaceColor', shiftColors{s}, 'MarkerEdgeColor', shiftColors{s});
    end
    for i = 1:numel(types)
        switch types{i}
            case 'stable'
                plot(pts(i,1), pts(i,2), 'p', 'MarkerSize', 7, 'MarkerFaceColor', black, 'MarkerEdgeColor', black);
            case 'unstable'
                plot(pts(i,1), pts(i,2), 'o', 'MarkerSize', 5, 'MarkerFaceColor', 'w', 'MarkerEdgeColor', black);
            otherwise
                plot(pts(i,1), pts(i,2), '+', 'MarkerSize', 8, 'LineWidth', 1.5, 'Color', black);
        end
    end
    text(-0.13, 1.04, char('A'+p-1), 'Units', 'normalized', 'FontSize', 11);
    text(0.05, 0.92, regimeLabels{p}, 'Units', 'normalized', 'FontSize', 9);
    xlabel('Re(z)'); if p == 1, ylabel('Im(z)'); end
    xlim([-1.02 1.02]); ylim([-1.02 1.02]);
end
legend('Location', 'southoutside', 'NumColumns', 4);
apply_kuramoto_style(fig);
exportgraphics(fig, fullfile(figDir, '15_circadian_phase_space_dynamics.png'), 'Resolution', 300, 'BackgroundColor', 'w');

%% Figure 16: recovery trajectories
Delta = 3.8e-3; K = 4.5*Delta; F = 3.5*Delta; Omega = 1.4*Delta;
[zst, ~, ~] = stable_fixed_point(K, F, Delta, Omega);
cases = [-3 -6 -9 12 3 6 9];
labels = {'3 E','6 E','9 E','12 E/W','3 W','6 W','9 W'};
cols = {blue, red, green, magenta, blue, red, green};
styles = {'-','-','-','-','--','--','--'};
fig = figure('Position', [100 100 470 335], 'Color', 'w'); hold on;
for i = 1:numel(cases)
    z0 = zst * exp(1i * cases(i) * 2*pi/24);
    [t, ztraj] = simulate_traj(z0, [0 14*24], K, F, Delta, Omega);
    dist = abs(ztraj - zst);
    idx = find(dist <= 0.2, 1, 'first');
    if isempty(idx), idx = numel(t); end
    plot(t(1:idx)/24, dist(1:idx), styles{i}, 'Color', cols{i}, 'LineWidth', 1.6, 'DisplayName', labels{i});
end
yline(0.2, 'Color', black, 'LineWidth', 1.1);
xlabel('Days'); ylabel('|z(t)-z_{st}|');
xlim([0 14]); ylim([0 2]); legend('Location', 'northeast', 'NumColumns', 2);
apply_kuramoto_style(fig);
exportgraphics(fig, fullfile(figDir, '16_circadian_recovery_trajectories.png'), 'Resolution', 300, 'BackgroundColor', 'w');

%% Figure 17: parameter dependence
Delta = 3.8e-3; Kref = 4.5*Delta; Fref = 3.5*Delta; Oref = 1.4*Delta;
scans = {linspace(2,15,40), linspace(1.5,5.8,40), linspace(-3.7,3.7,44)};
xlabels = {'K/\Delta','F/\Delta','\Omega/\Delta'};
ylims = {[2 24], [0 45], [0 34]};
fig = figure('Position', [100 100 820 280], 'Color', 'w');
for p = 1:3
    subplot(1,3,p); hold on;
    xs = scans{p};
    for c = 1:numel(cases)
        rec = nan(size(xs));
        for j = 1:numel(xs)
            switch p
                case 1
                    K = xs(j)*Delta; F = Fref; Omega = Oref;
                case 2
                    K = Kref; F = xs(j)*Delta; Omega = Oref;
                otherwise
                    K = Kref; F = Fref; Omega = xs(j)*Delta;
            end
            [zst, ~, ~] = stable_fixed_point(K, F, Delta, Omega);
            rec(j) = recovery_time(zst, cases(c), K, F, Delta, Omega);
        end
        plot(xs, rec, styles{c}, 'Color', cols{c}, 'LineWidth', 1.25, 'DisplayName', labels{c});
    end
    text(-0.15, 1.04, char('A'+p-1), 'Units', 'normalized', 'FontSize', 11);
    xlabel(xlabels{p}); if p == 1, ylabel('Recovery time (days)'); end
    ylim(ylims{p});
    if p == 2, legend('Location', 'northeast', 'NumColumns', 2, 'FontSize', 6.4); end
end
apply_kuramoto_style(fig);
exportgraphics(fig, fullfile(figDir, '17_circadian_parameter_dependence.png'), 'Resolution', 300, 'BackgroundColor', 'w');

function dz = complex_ode(z, K, F, Delta, Omega)
dz = 0.5*((K*z + F) - z.^2 .* (K*conj(z) + F)) - (Delta + 1i*Omega).*z;
end

function dy = ode_real(~, y, K, F, Delta, Omega)
z = y(1) + 1i*y(2);
dz = complex_ode(z, K, F, Delta, Omega);
dy = [real(dz); imag(dz)];
end

function [t, ztraj] = simulate_traj(z0, tspan, K, F, Delta, Omega)
opts = odeset('RelTol', 1e-8, 'AbsTol', 1e-10, 'MaxStep', 0.02);
if numel(tspan) == 2
    tspan = linspace(tspan(1), tspan(2), max(2500, ceil((tspan(2)-tspan(1))/0.02)+1));
end
[t, Y] = ode45(@(t,y) ode_real(t, y, K, F, Delta, Omega), tspan, [real(z0); imag(z0)], opts);
ztraj = Y(:,1) + 1i*Y(:,2);
end

function [zst, pts, types] = stable_fixed_point(K, F, Delta, Omega)
grid = linspace(-0.92, 0.92, 13);
pts = [];
opts = optimoptions('fsolve', 'Display', 'off', 'FunctionTolerance', 1e-12, 'StepTolerance', 1e-12);
for x = grid
    for y = grid
        if x^2 + y^2 <= 1
            try
                root = fsolve(@(yy) ode_real(0, yy, K, F, Delta, Omega), [x;y], opts);
                z = root(1) + 1i*root(2);
                if abs(z) <= 1.01
                    if isempty(pts) || all(vecnorm(pts - root', 2, 2) > 1e-4)
                        pts = [pts; root']; %#ok<AGROW>
                    end
                end
            catch
            end
        end
    end
end
types = cell(size(pts,1),1);
zst = pts(1,1) + 1i*pts(1,2);
for i = 1:size(pts,1)
    z = pts(i,1) + 1i*pts(i,2);
    J = numerical_jacobian(z, K, F, Delta, Omega);
    ev = eig(J);
    if all(real(ev) < 0)
        types{i} = 'stable';
        zst = z;
    elseif all(real(ev) > 0)
        types{i} = 'unstable';
    else
        types{i} = 'saddle';
    end
end
end

function J = numerical_jacobian(z, K, F, Delta, Omega)
epsv = 1e-5;
y = [real(z); imag(z)];
J = zeros(2,2);
for i = 1:2
    e = zeros(2,1); e(i) = epsv;
    J(:,i) = (ode_real(0, y+e, K, F, Delta, Omega) - ode_real(0, y-e, K, F, Delta, Omega)) / (2*epsv);
end
end

function rt = recovery_time(zst, shiftHours, K, F, Delta, Omega)
z0 = zst * exp(1i * shiftHours * 2*pi/24);
[t, ztraj] = simulate_traj(z0, [0 28*24], K, F, Delta, Omega);
dist = abs(ztraj - zst);
idx = find(dist <= 0.2, 1, 'first');
if isempty(idx)
    rt = NaN;
else
    rt = t(idx) / 24;
end
end

function cols = twilight_colors(n)
base = [ ...
    0.8858 0.8500 0.8879
    0.6443 0.4390 0.7389
    0.1849 0.0794 0.2131
    0.2453 0.2878 0.5373
    0.2890 0.5650 0.7595
    0.7590 0.7060 0.5680
    0.8858 0.8500 0.8879];
x = linspace(0, 1, size(base, 1));
xi = linspace(0.08, 0.88, n);
cols = interp1(x, base, xi);
end

File: Code/06_circadian_resynchronization/python/circadian_recovery.py
from __future__ import annotations

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from scipy.integrate import solve_ivp
from scipy.optimize import root

sys.path.append(str(Path(__file__).resolve().parents[2] / "common" / "python"))
from kuramoto_style import COLORS, clean_axes, panel_label, save_figure


def dzdt_complex(z, k, f, delta, omega):
    return 0.5 * ((k * z + f) - z**2 * (k * np.conj(z) + f)) - (delta + 1j * omega) * z


def rhs_real(_, y, k, f, delta, omega):
    z = y[0] + 1j * y[1]
    dz = dzdt_complex(z, k, f, delta, omega)
    return [dz.real, dz.imag]


def fixed_points(k, f, delta, omega):
    pts = []
    grid = np.linspace(-0.92, 0.92, 13)
    for x in grid:
        for y in grid:
            if x * x + y * y > 1:
                continue
            sol = root(lambda yy: rhs_real(0, yy, k, f, delta, omega), [x, y], method="hybr")
            if sol.success:
                z = sol.x[0] + 1j * sol.x[1]
                if abs(z) <= 1.01 and all(np.linalg.norm(sol.x - p) > 1e-4 for p in pts):
                    pts.append(sol.x)
    return np.asarray(pts)


def jacobian(z, k, f, delta, omega, eps=1e-5):
    y = np.array([z.real, z.imag])
    j = np.zeros((2, 2))
    for i in range(2):
        e = np.zeros(2)
        e[i] = eps
        j[:, i] = (np.asarray(rhs_real(0, y + e, k, f, delta, omega)) - np.asarray(rhs_real(0, y - e, k, f, delta, omega))) / (2 * eps)
    return j


def stable_point(k, f, delta, omega):
    pts = fixed_points(k, f, delta, omega)
    for p in pts:
        eig = np.linalg.eigvals(jacobian(p[0] + 1j * p[1], k, f, delta, omega))
        if np.all(np.real(eig) < 0):
            return p[0] + 1j * p[1], pts
    return pts[0, 0] + 1j * pts[0, 1], pts


def trajectory(z0, t_end, k, f, delta, omega, n=1800, max_step=np.inf):
    sol = solve_ivp(
        lambda t, y: rhs_real(t, y, k, f, delta, omega),
        (0, t_end),
        [z0.real, z0.imag],
        t_eval=np.linspace(0, t_end, n),
        rtol=1e-8,
        atol=1e-10,
        max_step=max_step,
    )
    return sol.t, sol.y[0] + 1j * sol.y[1]


def recovery_time(z_st, shift_hours, k, f, delta, omega, threshold=0.2, t_max_days=60):
    z0 = z_st * np.exp(1j * shift_hours * 2 * np.pi / 24)
    t, z = trajectory(z0, 24 * t_max_days, k, f, delta, omega, n=1200)
    d = np.abs(z - z_st)
    idx = np.where(d <= threshold)[0]
    return t[idx[0]] / 24 if len(idx) else np.nan


def phase_space():
    delta = 1.0
    f = 3.5 * delta
    omega = 1.4 * delta
    regimes = [(10.0 * delta, "$K=10\\Delta$"), (4.5 * delta, "$K=4.5\\Delta$")]
    twilight = plt.cm.twilight(np.linspace(0.08, 0.88, 4))
    shifts = [(-8.5, twilight[0], "8.5 h eastward"), (-9.5, twilight[1], "9.5 h eastward"), (9.0, twilight[2], "9.0 h westward"), (12.0, twilight[3], "12 h E/W")]

    fig, axes = plt.subplots(1, 2, figsize=(7.2, 3.35), sharex=True, sharey=True)
    theta = np.linspace(0, 2 * np.pi, 500)
    for i, (k, label) in enumerate(regimes):
        ax = axes[i]
        z_st, pts = stable_point(k, f, delta, omega)
        ax.plot(np.cos(theta), np.sin(theta), "--", lw=1.0, color=COLORS["light_gray"])
        ax.axhline(0, color=COLORS["light_gray"], lw=0.9)
        ax.axvline(0, color=COLORS["light_gray"], lw=0.9)
        for radius in [0.25, 0.55, 0.85]:
            for angle in np.linspace(0, 2 * np.pi, 18, endpoint=False):
                z0 = radius * np.exp(1j * angle)
                _, z_flow = trajectory(z0, 25, k, f, delta, omega, n=1200, max_step=0.035)
                ax.plot(z_flow.real, z_flow.imag, color=COLORS["light_gray"], lw=0.55, alpha=0.50, zorder=0)
        for hours, color, text in shifts:
            z0 = z_st * np.exp(1j * hours * 2 * np.pi / 24)
            _, z = trajectory(z0, 80, k, f, delta, omega, n=2600, max_step=0.035)
            ax.plot(z.real, z.imag, lw=1.5, color=color, label=text)
            ax.plot(z.real[0], z.imag[0], "o", ms=3.2, color=color)
        for p in pts:
            z = p[0] + 1j * p[1]
            eig = np.linalg.eigvals(jacobian(z, k, f, delta, omega))
            if np.all(np.real(eig) < 0):
                ax.plot(p[0], p[1], "p", ms=7, color=COLORS["black"])
            elif np.all(np.real(eig) > 0):
                ax.plot(p[0], p[1], "o", ms=5, mfc="white", mec=COLORS["black"])
            else:
                ax.plot(p[0], p[1], "+", ms=8, mew=1.5, color=COLORS["black"])
        panel_label(ax, chr(ord("A") + i), -0.13, 1.04)
        ax.text(0.05, 0.92, label, transform=ax.transAxes, fontsize=9)
        ax.set_aspect("equal")
        ax.set_xlim(-1.02, 1.02)
        ax.set_ylim(-1.02, 1.02)
        clean_axes(ax, "$\\operatorname{Re}(z)$", "$\\operatorname{Im}(z)$" if i == 0 else None)
    axes[0].legend(frameon=False, loc="lower center", bbox_to_anchor=(1.04, -0.33), ncol=4)
    save_figure(fig, "15_circadian_phase_space_dynamics.png")


def recovery_curves():
    delta = 3.8e-3
    k, f, omega = 4.5 * delta, 3.5 * delta, 1.4 * delta
    z_st, _ = stable_point(k, f, delta, omega)
    twilight = plt.cm.twilight(np.linspace(0.08, 0.88, 4))
    cases = [(-3, twilight[0], "3 E", "-"), (-6, twilight[1], "6 E", "-"), (-9, twilight[2], "9 E", "-"), (12, twilight[3], "12 E/W", "-"), (3, twilight[0], "3 W", "--"), (6, twilight[1], "6 W", "--"), (9, twilight[2], "9 W", "--")]
    fig, ax = plt.subplots(figsize=(4.7, 3.35))
    for hours, color, label, ls in cases:
        z0 = z_st * np.exp(1j * hours * 2 * np.pi / 24)
        t, z = trajectory(z0, 14 * 24, k, f, delta, omega, n=900)
        d = np.abs(z - z_st)
        keep = np.arange(len(d))
        reached = np.where(d <= 0.2)[0]
        if len(reached):
            keep = keep[: reached[0] + 1]
        ax.plot(t[keep] / 24, d[keep], ls=ls, lw=1.6, color=color, label=label)
    ax.axhline(0.2, color=COLORS["black"], lw=1.1)
    ax.set_xlim(0, 14)
    ax.set_ylim(0, 2.0)
    clean_axes(ax, "Days", "$|z(t)-z_{st}|$")
    ax.legend(frameon=False, loc="upper right", ncol=2)
    save_figure(fig, "16_circadian_recovery_trajectories.png")


def parameter_dependence():
    delta = 3.8e-3
    k_ref, f_ref, omega_ref = 4.5 * delta, 3.5 * delta, 1.4 * delta
    twilight = plt.cm.twilight(np.linspace(0.08, 0.88, 4))
    cases = [(-3, twilight[0], "3 E", "-"), (-6, twilight[1], "6 E", "-"), (-9, twilight[2], "9 E", "-"), (12, twilight[3], "12 E/W", "-"), (3, twilight[0], "3 W", "--"), (6, twilight[1], "6 W", "--"), (9, twilight[2], "9 W", "--")]
    scans = [
        ("$K/\\Delta$", np.linspace(2.0, 15.0, 72), lambda x: (x * delta, f_ref, omega_ref), (2, 24)),
        ("$F/\\Delta$", np.linspace(1.5, 5.8, 70), lambda x: (k_ref, x * delta, omega_ref), (0, 45)),
        ("$\\Omega/\\Delta$", np.linspace(-3.7, 3.7, 78), lambda x: (k_ref, f_ref, x * delta), (0, 34)),
    ]
    fig, axes = plt.subplots(1, 3, figsize=(8.2, 2.8))
    for j, (xlabel, xs, params, ylim) in enumerate(scans):
        ax = axes[j]
        for shift, color, label, ls in cases:
            rec = []
            for x in xs:
                k, f, omega = params(x)
                z_st, _ = stable_point(k, f, delta, omega)
                rec.append(recovery_time(z_st, shift, k, f, delta, omega, threshold=0.2, t_max_days=28))
            ax.plot(xs, rec, ls=ls, lw=1.25, color=color, label=label)
        panel_label(ax, chr(ord("A") + j), -0.15, 1.04)
        ax.set_ylim(*ylim)
        clean_axes(ax, xlabel, "Recovery time (days)" if j == 0 else None)
        if j == 1:
            ax.legend(frameon=False, loc="upper right", fontsize=6.4, ncol=2)
    fig.subplots_adjust(wspace=0.28)
    save_figure(fig, "17_circadian_parameter_dependence.png")


if __name__ == "__main__":
    phase_space()
    recovery_curves()
    parameter_dependence()
